### PR TITLE
Generate serializer for type arguments of IAsyncObserver<T>

### PR DESCRIPTION
**Pull request by dVakulen**
#1316

### CODE REVIEW

File: src/TesterInternal/SerializerTests/SerializerGenerationTests.cs [hunk 1]
```diff
@@ -1,32 +0,0 @@
...
-using Orleans.Serialization;
-using Orleans.UnitTest.GrainInterfaces;
-//using Orleans.UnitTest.Unsigned.GrainInterfaces;
-
-//using UnitTestGrains;
-//using Echo.Grains;
-
-// ReSharper disable NotAccessedVariable
-
-namespace UnitTests.SerializerTests
```
Comment (dVakulen) [inline src/TesterInternal/SerializerTests/SerializerGenerationTests.cs]: Removed:
https://github.com/dotnet/orleans/blob/master/src/TesterInternal/SerializerTests/SerializerGenerationTests.cs
https://github.com/dotnet/orleans/blob/master/src/TesterInternal/Serialization/SerializerGenerationTests.cs

**Comment by sergeybykov** [comment]
This was almost too fast, @dVakulen! :-) 
LGTM but I'll wait for @ReubenBond's word before merging it.

**Comment by gabikliot** [comment]
It's good, but we also need `IAsyncObservable`, and `IAsyncStream`, and `StreamSubscriptionHandle`.

File: src/TesterInternal/SerializerTests/CustomSerializerTests.cs [hunk 2]
```diff
@@ -132,5 +133,13 @@ public void Serialize_GrainMethodAsyncReturnOnlyType()
         {
             Assert.IsNotNull(SerializationManager.GetSerializer(typeof(SerializerTestClass6)), "No serializer generated for return type of parameterless Task grain method");
         }
+        
+        [TestMethod, TestCategory("Functional"), TestCategory("Serialization")]
+        public void Serialize_AsyncObserverArgumentType()
+        {
+            var z = new ClosedTypeStreamObserver(TraceLogger.GetLogger("SerializationTests", TraceLogger.LoggerType.Runtime));
+
+            Assert.IsNotNull(SerializationManager.GetSerializer(typeof(AsyncObserverArg)), "No serializer generated for argument type of async observer");
```
Comment (gabikliot) [inline src/TesterInternal/SerializerTests/CustomSerializerTests.cs]: Maybe also worth trying to actually serialize it as well? make sure it does not throw?

File: src/TesterInternal/SerializerTests/CustomSerializerTests.cs [hunk 3]
```diff
@@ -132,5 +133,13 @@ public void Serialize_GrainMethodAsyncReturnOnlyType()
         {
             Assert.IsNotNull(SerializationManager.GetSerializer(typeof(SerializerTestClass6)), "No serializer generated for return type of parameterless Task grain method");
         }
+        
+        [TestMethod, TestCategory("Functional"), TestCategory("Serialization")]
+        public void Serialize_AsyncObserverArgumentType()
+        {
+            var z = new ClosedTypeStreamObserver(TraceLogger.GetLogger("SerializationTests", TraceLogger.LoggerType.Runtime));
```
Comment (gabikliot) [inline src/TesterInternal/SerializerTests/CustomSerializerTests.cs]: why do you need to instantiate `ClosedTypeStreamObserver` here?

File: src/OrleansCodeGenerator/RoslynCodeGenerator.cs [hunk 4]
```diff
@@ -429,10 +429,13 @@ private static GeneratedSyntax GenerateForAssemblies(List<Assembly> assemblies,
...
-            Type grainStateType;
+            Type typeToRecord;
             // If a type extends Grain<T>, add T to the list of candidates for serialization
-            if (IsPersistentGrain(typeInfo, out grainStateType))
-                RecordType(grainStateType, module, targetAssembly, includedTypes);
+            if (IsPersistentGrain(typeInfo, out typeToRecord))
+                RecordType(typeToRecord, module, targetAssembly, includedTypes);
+
+            if (IsAsyncObserver(typeInfo, out typeToRecord))
+                RecordType(typeToRecord, module, targetAssembly, includedTypes);
```
Comment (ReubenBond) [inline src/OrleansCodeGenerator/RoslynCodeGenerator.cs]: We could also remove the condition and call `RecordType` for all types.

**Comment by ReubenBond** [comment]
LGTM after addressing the comments, which are minor. I made a comment about removing the conditional checks and always recording the type. We only consider a subset of assemblies now anyhow, so it might not be a significant performance hit and it will mean less special-casing.

File: src/OrleansCodeGenerator/RoslynCodeGenerator.cs (hunk 4 above)
Comment (gabikliot) [inline src/OrleansCodeGenerator/RoslynCodeGenerator.cs]: You mean record ALL types? But that would mean ANY type defined in the assembly, including ones that were never intended to be serialized. Would not that be a problem? Would we always succeed, on any type?

Comment (ReubenBond) [inline src/OrleansCodeGenerator/RoslynCodeGenerator.cs]: Record really just means "record if serializable", it performs accessibility checks first

File: src/TesterInternal/SerializerTests/CustomSerializerTests.cs (hunk 2 above)
Comment (dVakulen) [inline src/TesterInternal/SerializerTests/CustomSerializerTests.cs]: That would be rather sanity check and test of a generated serializer, but also having that basically does not hurt, so I've added additional assert.

File: src/TesterInternal/SerializerTests/CustomSerializerTests.cs (hunk 3 above)
Comment (dVakulen) [inline src/TesterInternal/SerializerTests/CustomSerializerTests.cs]: Removed

**Comment by dVakulen** [comment]
Comments addressed, always recording the type won't fit in this PR as there were some issues after enabling this behavior: 
 - bug with generation of setters\getters of fields with private nested type;
 - serializers for ```<PrivateImplementationDetails>```;
 - resulting generated code became ~4x bigger, which could lead to longer startup times and constantly higher memory consumption by ```SerializationManager```.

While first two issues will be fixed in the near future, the latest one needs more weighted consideration of all its pros and cons.

File: src/OrleansCodeGenerator/RoslynCodeGenerator.cs [hunk 5]
```diff
@@ -449,16 +447,35 @@ private static void RecordType(Type type, Module module, Assembly targetAssembly
...
-        private static bool IsPersistentGrain(TypeInfo typeInfo, out Type stateType)
+        private static void ConsiderGenericBaseTypeArguments(
+            TypeInfo typeInfo,
+            Module module,
+            Assembly targetAssembly,
+            ISet<Type> includedTypes)
         {
-            stateType = null;
+            if (typeInfo.BaseType == null) return;
+            if (!typeInfo.BaseType.IsGenericType) return;
```
Comment (ReubenBond) [inline src/OrleansCodeGenerator/RoslynCodeGenerator.cs]: should this be IsConstructedGenericType & likewise below?

**Comment by sergeybykov** [comment]
I think this is already a significant improvement as is. I agree with @dVakulen that the question of always recording the type can be explored separately.

**Comment by ReubenBond** [comment]
Thanks, Dmitry! One last comment, LGTM either way

File: src/TestInternalGrains/StreamLifecycleTestGrains.cs [hunk 6]
```diff
@@ -34,6 +34,54 @@ public StreamLifecycleTestGrainState()
...
+            {
+                return false;
+            }
+
+            return A.Equals(item.A) && B.Equals(item.B);
+        }
+
+        public override int GetHashCode()
+        {
+            return B ^ A.GetHashCode();
```
Comment (veikkoeeva) [inline src/TestInternalGrains/StreamLifecycleTestGrains.cs]: I'm very tired, so apologizes if this is false memory, but I have a faint recall of problems of this sort: http://stackoverflow.com/questions/892618/create-a-hashcode-of-two-numbers. Do they apply?

**Comment by veikkoeeva** [comment]
I addded a comments just before dropping to the keyboard. Maybe a false alarm, as you guys and @dVakulen do a stellar job. :)

File: src/TestInternalGrains/StreamLifecycleTestGrains.cs (hunk 6 above)
Comment (ReubenBond) [inline src/TestInternalGrains/StreamLifecycleTestGrains.cs]: ReSharper generates this implementation:
```C#
return (this.B * 397) ^ (this.A != null ? this.A.GetHashCode() : 0);
```

File: src/OrleansCodeGenerator/RoslynCodeGenerator.cs (hunk 5 above)
Comment (dVakulen) [inline src/OrleansCodeGenerator/RoslynCodeGenerator.cs]: Fixed.

File: src/TestInternalGrains/StreamLifecycleTestGrains.cs (hunk 6 above)
Comment (dVakulen) [inline src/TestInternalGrains/StreamLifecycleTestGrains.cs]: I don't think that http://stackoverflow.com/questions/892618/create-a-hashcode-of-two-numbers applies to this case; Changed to the ReSharper's implementation.

**Comment by gabikliot** [comment]
Thanks @dVakulen ! Spot on, as usual.

**Comment by sergeybykov** [comment]
:+1:

**Pull request merged**